app/ante/eip712.go

```diff
@@ -0,0 +1,292 @@
+
+func init() {
+	registry := codectypes.NewInterfaceRegistry()
+	ethermint.RegisterInterfaces(registry)
+	ethermintCodec = codec.NewProtoCodec(registry)
+}
+
+// Verify all signatures for a tx and return an error if any are invalid. Note,
+// the Eip712SigVerificationDecorator decorator will not get executed on ReCheck.
+//
+// CONTRACT: Pubkeys are set in context for all signers before this decorator runs
+// CONTRACT: Tx must implement SigVerifiableTx interface
+type Eip712SigVerificationDecorator struct {
+	ak              evmtypes.AccountKeeper
+	signModeHandler authsigning.SignModeHandler
+}
+
+func NewEip712SigVerificationDecorator(ak evmtypes.AccountKeeper, signModeHandler authsigning.SignModeHandler) Eip712SigVerificationDecorator {
+	return Eip712SigVerificationDecorator{
+		ak:              ak,
+		signModeHandler: signModeHandler,
+	}
+}
+
+func (svd Eip712SigVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	// no need to verify signatures on recheck tx
+	if ctx.IsReCheckTx() {
+		return next(ctx, tx, simulate)
+	}
+
+	sigTx, ok := tx.(authsigning.SigVerifiableTx)
+	if !ok {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "tx %T doesn't implement authsigning.SigVerifiableTx", tx)
+	}
+
+	// stdSigs contains the sequence number, account number, and signatures.
+	// When simulating, this would just be a 0-length slice.
+	sigs, err := sigTx.GetSignaturesV2()
+	if err != nil {
+		return ctx, err
+	}
+
+	signerAddrs := sigTx.GetSigners()
+
+	// EIP712 allows just one signature
+	if len(sigs) != 1 {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "invalid number of signer;  EIP712 signatures allows just one signature")
+	}
+
+	// check that signer length and signature length are the same
+	if len(sigs) != len(signerAddrs) {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "invalid number of signer;  expected: %d, got %d", len(signerAddrs), len(sigs))
+	}
+
+	// EIP712 has just one signature, avoid looping here and only read index 0
+	i := 0
+	sig := sigs[i]
+
+	acc, err := authante.GetSignerAcc(ctx, svd.ak, signerAddrs[i])
+	if err != nil {
+		return ctx, err
+	}
+
+	// retrieve pubkey
+	pubKey := acc.GetPubKey()
+	if !simulate && pubKey == nil {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidPubKey, "pubkey on account is not set")
+	}
+
+	// Check account sequence number.
+	if sig.Sequence != acc.GetSequence() {
+		return ctx, sdkerrors.Wrapf(
+			sdkerrors.ErrWrongSequence,
+			"account sequence mismatch, expected %d, got %d", acc.GetSequence(), sig.Sequence,
+		)
+	}
+
+	// retrieve signer data
+	genesis := ctx.BlockHeight() == 0
+	chainID := ctx.ChainID()
+
+	var accNum uint64
+	if !genesis {
+		accNum = acc.GetAccountNumber()
+	}
+
+	signerData := authsigning.SignerData{
+		ChainID:       chainID,
+		AccountNumber: accNum,
+		Sequence:      acc.GetSequence(),
+	}
+
+	if simulate {
+		return next(ctx, tx, simulate)
```

Comment:
Here I just removed the loop (because we have just one signature) and call next if we are simulating